File: dsaWithJava/tree/binarysearchtree/IsBST.java
package tree.binarysearchtree;

public class IsBST {
    // Time Complexity = O(n)
    public boolean isBST(Node root) {
    return isBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean isBST(Node root, long min, long max) {
        if (root == null)
            return true;
        if (root.data <= min || root.data >= max)
            return false;
        return (isBST(root.left, min, root.data) &&
                isBST(root.right, root.data, max));
    }
    /*
    int prev = Integer.MIN_VALUE;

    public boolean isBST(Node root) {
        if (root == null)
            return true;
        if (isBST(root.left) == false)
            return false;
        if (root.data <= prev)
            return false;
        prev = root.data;
        return isBST(root.right);
    }
    */
}

File: Python/tree/binary_search_tree/is_bst.py
import sys
from node import Node


def is_bst(root: Node) -> bool:
    def __is_bst(root: Node, min_, max_):
        if not root:
            return True
        if root.data <= min_ or root.data >= max_:
            return False
        return __is_bst(root.left, min_, root.data) and __is_bst(
            root.right, root.data, max_
        )

    return __is_bst(root, -sys.maxsize, sys.maxsize)


# prev = -sys.maxsize


# def is_bst(root: Node) -> bool:
#     if not root:
#         return True
#     if is_bst(root.left) == False:
#         return False
#     if root.data <= prev:
#         return False
#     prev = root.data
#     return is_bst(root.right)
